usr/sbin/p11sak/p11sak.c

```diff
@@ -1141,6 +1191,200 @@ static CK_ATTRIBUTE_TYPE col2type(int col)
+        return CKR_ARGUMENTS_BAD;
+    }
+    if (parse_configlib_file(fp, &cfg, error_hook, 0)) {
+        fprintf(stderr, "\nFailed to parse %s\n", file_loc);
+        return CKR_DATA_INVALID;
+    }
+    fclose(fp);
+
+    // check if at least one attribute is defined
+    if (cfg != NULL) {
+        confignode_foreach(c, cfg, f) {
+            if (confignode_hastype(c, CT_STRUCT) && strcmp(c->key, "attribute") == 0) {
+                structnode = confignode_to_struct(c);
+                name = confignode_find(structnode->value, "name");
+                hex_string = confignode_find(structnode->value, "id");
+                type = confignode_find(structnode->value, "type");
+
+                // checking syntax of attribute...
+                if (confignode_hastype(name, CT_BAREVAL) && name != NULL) {
+                    if (confignode_hastype(hex_string, CT_INTVAL) && hex_string != NULL) {
+                        if (confignode_hastype(type, CT_BAREVAL) && type != NULL) {
+                            printf(" 1 ");
+                            return rc;
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    printf(" 0 ");
+    return rc;
+}
+
+/**
+ *  Print in p11sak_defined_attrs.conf defined attributes in long format
+ */
+static CK_RV long_print_vendor(CK_SESSION_HANDLE session, CK_OBJECT_HANDLE hkey) 
+{
+    CK_RV rc = CKR_OK;
+    int f;
+    FILE *fp;
+    struct ConfigBaseNode *cfg, *c, *name, *hex_string, *type;
+    struct ConfigStructNode *structnode;
+    char file_loc[1024] = P11SAK_DEFINED_ATTRS_LOCATION;
+
+    // get file location
+    cfg = NULL;
+    if (strcmp(P11SAK_DEFINED_ATTRS_LOCATION, "") == 1) {
+        strcpy(file_loc, P11SAK_DEFINED_ATTRS_LOCATION);
+    } else {
+        strcpy(file_loc, P11SAK_DEFAULT_CONF_FILE);
+    } 
+
+    // open and parse file
+    if ((fp = fopen(file_loc, "r")) == NULL) {
+        fprintf(stderr, "Failed to load %s: %s\n", file_loc, strerror(errno));
+        return CKR_ARGUMENTS_BAD;
+    }
+    if (parse_configlib_file(fp, &cfg, error_hook, 0)) {
+        fprintf(stderr, "Failed to parse %s\n", file_loc);
+        return CKR_DATA_INVALID;
+    }
+
+    // parse and print attributes
+    fclose(fp);
+    if (cfg != NULL)
+    {
+        confignode_foreach(c, cfg, f) {
+            if (confignode_hastype(c, CT_STRUCT) && strcmp(c->key, "attribute") == 0) {
+                // parse...
+                structnode = confignode_to_struct(c);
+                name = confignode_find(structnode->value, "name");
+                hex_string = confignode_find(structnode->value, "id");
+                type = confignode_find(structnode->value, "type");
+
+                // checking syntax of attribute...
+                if (name == NULL || !confignode_hastype(name, CT_BAREVAL)) {
+                    fprintf(stderr, "Invalid name in Attribute in line %hu\n", c->line);
```

Comment:
`c` is the structure.  So its line might not be the right line here.  I would split the `if` into two cases:
1. no `name` in the structure => error: "Missing name in structure at line %hu\n", c->line
2. `name` has wrong type => error: "Invalid 'name' attribute in line %lu\n", name->line
And similarly for the other two.

Reply:
Done

Reply:
No, here you still use the line of the structure and not that of the name.